src/core/Akka/IO/UdpListener.cs

```diff
@@ -89,46 +84,41 @@ protected override bool Receive(object message)
 
         private bool ReadHandlers(object message)
         {
-            if (message is SuspendReading)
-            {
-                // TODO: What should we do here - we cant cancel a pending ReceiveAsync
-                return true;
-            }
-            if (message is ResumeReading)
-            {
-                ReceiveAsync();
-                return true;
-            }
-            if (message is SocketReceived)
+            switch (message)
             {
-                var received = (SocketReceived) message;
-                DoReceive(received.EventArgs, _bind.Handler);
-                return true;
+                case SuspendReading _:
+                    // TODO: What should we do here - we cant cancel a pending ReceiveAsync
+                    return true;
+                case ResumeReading _:
+                    ReceiveAsync();
+                    return true;
+                case SocketReceived _:
+                    var received = (SocketReceived) message;
+                    DoReceive(received.EventArgs, _bind.Handler);
+                    return true;
+                case Unbind _:
+                    Log.Debug("Unbinding endpoint [{0}]", _bind.LocalAddress);
+                    try
+                    {
+                        Socket.Dispose();
+                        Sender.Tell(Unbound.Instance);
+                        Log.Debug("Unbound endpoint [{0}], stopping listener", _bind.LocalAddress);
+                    }
+                    finally
+                    {
+                        Context.Stop(Self);
+                    }
+                    return true;
             }
 
-            if (message is Unbind)
-            {
-                Log.Debug("Unbinding endpoint [{0}]", _bind.LocalAddress);
-                try
-                {
-                    Socket.Dispose();
-                    Sender.Tell(Unbound.Instance);
-                    Log.Debug("Unbound endpoint [{0}], stopping listener", _bind.LocalAddress);
-                }
-                finally
-                {
-                    Context.Stop(Self);
-                }
-                return true;
-            }
             return false;
         }
 
         private void DoReceive(SocketAsyncEventArgs e, IActorRef handler)
         {
             try
             {
-                handler.Tell(new Received(ByteString.FromBytes(e.Buffer, e.Offset, e.BytesTransferred), e.RemoteEndPoint));
+                handler.Tell(new Received(ByteString.CopyFrom(e.Buffer, e.Offset, e.BytesTransferred), e.RemoteEndPoint));
```

Comment:
The bug itself was here - we were creating `ByteString` directly from buffer assigned to a socket and the we continued to receive more bytes (no copying). When they arrived, they were overriding internal content of byte string we already passed to the user.